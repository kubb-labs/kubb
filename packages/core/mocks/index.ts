import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
