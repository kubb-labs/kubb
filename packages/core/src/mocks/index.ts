import path from 'node:path'
import { getSource } from '../FileManager'
import type { File, ResolvedFile } from '../fs/types.ts'
import type { Logger } from '../logger'
import type { PluginManager } from '../PluginManager.ts'
import { camelCase, pascalCase } from '../transformers/casing.ts'
import type { Plugin } from '../types.ts'
import { createFile } from '../utils'

export const mockedLogger = {
  emit(_type, _message) {},
  on(_type, _message) {},
  consola: {},
} as Logger

export const createMockedPluginManager = (name?: string) =>
  ({
    resolveName: (result) => {
      if (result.type === 'file') {
        return camelCase(name || result.name)
      }

      if (result.type === 'type') {
        return pascalCase(result.name)
      }

      if (result.type === 'function') {
        return camelCase(result.name)
      }

      return camelCase(result.name)
    },
    config: {
      output: {
        path: './path',
      },
    },
    resolvePath: ({ baseName }) => baseName,
    logger: {
      emit(message) {
        console.log(message)
      },
      on(_eventName, _args) {},
      logLevel: 3,
    },
    getPluginByKey: (_pluginKey: Plugin['key']) => {
      return undefined
    },
    getFile: ({ name, extname, pluginKey }) => {
      const baseName = `${name}${extname}`

      return {
        path: baseName,
        baseName,
        meta: {
          pluginKey,
        },
      }
    },
  }) as PluginManager

export const mockedPluginManager = createMockedPluginManager('')

export async function matchFiles(files: Array<ResolvedFile | File> | undefined, pre?: string) {
  if (!files) {
    return undefined
  }

  for await (const file of files) {
    const source = await getSource(createFile(file), { logger: mockedLogger })
    await expect(source).toMatchFileSnapshot(path.join(...['__snapshots__', pre, file.path].filter(Boolean)))
  }
}
