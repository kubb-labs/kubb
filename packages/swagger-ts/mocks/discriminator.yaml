info:
  title: Swagger Petstore
openapi: 3.1.0
components:
  schemas:
    Cat:
      title: Cat
      type: object
      properties:
        id:
          format: uuid
          title: Id
          type: string
        type:
          const: Cat
          title: Type
        meow:
          type: boolean
          title: Meowing prevalence
      required:
      - id
      - type
    Dog:
      title: Dog
      type: object
      properties:
        id:
          format: uuid
          title: Id
          type: string
        bark:
          type: number
          title: Barking frequency
        type:
          const: Dog
          title: Type
      required:
      - id
      - type
    Petstore:
      title: Petstore
      type: object
      description: A project is an analysis project. It has a name, an id, and has
        a dataset
      properties:
        id:
          title: Id
          type: integer
        pets:
          default: []
          items:
            discriminator:
              mapping:
                Cat: "#/components/schemas/Cat"
                Dog: "#/components/schemas/Dog"
              propertyName: type
            oneOf:
            - "$ref": "#/components/schemas/Cat"
            - "$ref": "#/components/schemas/Dog"
          title: Pets
          type: array
      required:
      - id

paths:
