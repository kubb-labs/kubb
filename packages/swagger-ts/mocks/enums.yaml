
definitions:
  enumVarNames.Type:
    enum:
    - 0
    - 1
    type: integer
    x-enum-varnames:
    - Pending
    - Received
  enumNames.Type:
    enum:
    - 0
    - 1
    x-enumNames:
      - Pending
      - Received
  enum.String:
    type: string
    enum:
      - created_at
      - description
      - FILE.UPLOADED
      - FILE.DOWNLOADED
  enum.InObject:
    type: object
    properties:
      reason:
        type: string
        enum:
          - created_at
          - description
  enum.AllOf:
    type: object
    properties:
      reason:
        allOf:
          - type: string
            enum:
              - created_at
              - description

  enum.Array:
    type: object
    properties:
     identifier:
        type: array
        prefixItems:
          - type: integer
            description: Number of items
          - type: string
            description: Description
          - enum:
              - "NW"
              - "NE"
              - "SW"
              - "SE"
            description: Wind direction

  enum.Items:
    type: array
    minItems: 1
    uniqueItems: true
    items:
      type: string
      enum:
        - created_at
        - description
swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths:
  /var-names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumVarNames.Type'
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/enumNames.Type'
