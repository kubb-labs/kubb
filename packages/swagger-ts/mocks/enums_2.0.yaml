definitions:
  enum.Array:
    type: object
    properties:
      identifier:
        type: array
        prefixItems:
          - type: integer
            description: Number of items
          - type: string
            description: Description
          - enum:
              - "NW"
              - "NE"
              - "SW"
              - "SE"
            description: Wind direction

swagger: "2.0"
info:
  title: test
  version: "1.0.0"
paths: {}
