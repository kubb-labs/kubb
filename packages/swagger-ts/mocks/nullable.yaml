
definitions:
  Nullable:
    type: object
    properties:
      foo:
        type: string
        nullable: true
openapi: 3.1.0
info:
  title: test
  version: "1.0.0"
paths:
  /names:
    get:
      consumes:
      - application/json
      responses:
        200:
          description: Success
          schema:
            $ref: '#/definitions/Nullable'
