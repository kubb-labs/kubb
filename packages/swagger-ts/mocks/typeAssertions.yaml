info:
  title: Type Assertions
openapi: 3.1.0
paths: {}
components:
  schemas:
    Body_upload_file_api_assets_post:
      properties:
        file:
          format: binary
          title: File
          type: string
      required:
      - file
      title: Body_upload_file_api_assets_post
      type: object
      $comment: |
        export type BodyUploadFileApiAssetsPost = {
            /**
            * @type string binary
            */
            file: Blob;
        };
    Plain_file:
      format: binary
      title: File
      type: string
      $comment: export type PlainFile = Blob;
    Plain_date:
      format: date
      title: Date
      type: string
      $comment: export type PlainDate = Date;
    Plain_time:
      format: time
      title: Time
      type: number
      $comment: export type PlainTime = number;
    Plain_date_time:
      format: date-time
      title: Datetime
      type: string
      $comment:
    Plain_email:
      format: email
      title: Email
      type: string
    Plain_uuid:
      format: uuid
      title: UUID
      type: string

