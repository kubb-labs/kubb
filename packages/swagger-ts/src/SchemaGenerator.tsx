import { createRoot } from '@kubb/react'
import { SchemaGenerator as Generator } from '@kubb/swagger'
import { Oas, Schema } from '@kubb/swagger/components'
import { pluginKey as swaggerTypeScriptPluginKey } from '@kubb/swagger-ts'

import { pluginKey } from './plugin.ts'
import { typeParser } from './typeParser.ts'

import type { AppContextProps } from '@kubb/react'
import type { Schema as SchemaType, SchemaGeneratorBuildOptions, SchemaMethodResult } from '@kubb/swagger'
import type { SchemaObject } from '@kubb/swagger/oas'
import type { FileMeta, PluginOptions } from './types.ts'

export class SchemaGenerator extends Generator<PluginOptions['resolvedOptions'], PluginOptions> {
  async schema(name: string, object: SchemaObject): SchemaMethodResult<FileMeta> {
    const { oas, pluginManager, mode, plugin, output } = this.context

    const root = createRoot<AppContextProps<PluginOptions['appMeta']>>({
      logger: pluginManager.logger,
    })

    root.render(
      <Oas oas={oas}>
        <Oas.Schema generator={this} name={name} object={object}>
          <Schema.File isTypeOnly output={output} mode={mode} />
        </Oas.Schema>
      </Oas>,
      { meta: { pluginManager, plugin } },
    )

    return root.files
  }
  // TODO convert to a react component called `Schema.Parser` with props parser as part of the SchemaContext
  getSource(name: string, schemas: SchemaType[], { keysToOmit, description }: SchemaGeneratorBuildOptions = {}): string[] {
    const texts: string[] = []

    const resolvedName = this.context.pluginManager.resolveName({
      name,
      pluginKey,
      type: 'function',
    })
    const resvoledTypeName = this.context.pluginManager.resolveName({
      name,
      pluginKey: swaggerTypeScriptPluginKey,
      type: 'type',
    })

    const typeOutput = typeParser(schemas, {
      name: resolvedName,
      typeName: resvoledTypeName,
      description,
      enumType: this.options.enumType || 'asConst',
      optionalType: this.options.optionalType,
      keysToOmit,
    })

    texts.push(typeOutput)

    return texts
  }
  /**
   * @deprecated only used for testing
   */

  buildSource(name: string, schema: SchemaObject, options: SchemaGeneratorBuildOptions = {}): string[] {
    const schemas = this.buildSchemas(schema, name)

    return this.getSource(name, schemas, options)
  }
}
