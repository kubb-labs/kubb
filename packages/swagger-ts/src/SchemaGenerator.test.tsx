import path from 'node:path'

import { mockedPluginManager } from '@kubb/core/mocks'
import { OasManager } from '@kubb/swagger'

import { SchemaGenerator } from './SchemaGenerator.tsx'

import type { Plugin, PluginManager } from '@kubb/core'
import type { Oas, OasTypes, SchemaObject } from '@kubb/swagger/oas'
import type { PluginOptions } from './types.ts'

describe('TypeScript SchemaGenerator petStore', async () => {
  const petStorePath = path.resolve(__dirname, '../mocks/petStore.yaml')
  const oas = await new OasManager().parse(petStorePath)

  test('generate type for Pet with optionalType `questionToken`', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: 'enum',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pet', schemas?.Pet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate type for Pet with optionalType `undefined`', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'undefined',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pet', schemas?.Pet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate type for Pet with optionalType `questionTokenAndUndefined`', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionTokenAndUndefined',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pet', schemas?.Pet as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate type for nullable fields', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas: {} as Oas,
        pluginManager: {
          resolveName: ({ name }) => name,
          resolvePath: ({ baseName }) => baseName,
        } as PluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schema: OasTypes.SchemaObject = {
      type: 'object',
      properties: {
        foo: {
          type: 'string',
          nullable: true,
        },
      },
    }

    const node = generator.buildSource('Test', schema)

    expect(node).toMatchSnapshot()
  })

  test('generate type for Pets', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pets', schemas?.Pets as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('SchemaGenerator petStoreRef', async () => {
  const petStoreRefPath = path.resolve(__dirname, '../mocks/petStoreRef.yaml')
  const oas = await new OasManager().parse(petStoreRefPath)

  test('generate type for Pets', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Pets', schemas?.Pets as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('SchemaGenerator discriminator', async () => {
  const discriminatorPath = path.resolve(__dirname, '../mocks/discriminator.yaml')
  const oas = await new OasManager().parse(discriminatorPath)

  test('PetStore defined as array with type union', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Petstore', schemas?.Petstore as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('Cat.type defined as const', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Cat', schemas?.Cat as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('Dog.type defined as const', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('Dog', schemas?.Dog as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('NullConst correctly produces "null"', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )
    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('NullConst', schemas?.NullConst as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('StringValueConst const correctly produces "foobar"', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('StringValueConst', schemas?.StringValueConst as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('NumberValueConst const correctly produces `42`', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('NumberValueConst', schemas?.NumberValueConst as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('MixedValueTypeConst ignores type constraint in favor of const constraint', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const schemas = oas.getDefinition().components?.schemas
    const node = generator.buildSource('MixedValueTypeConst', schemas?.MixedValueTypeConst as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})

describe('SchemaGenerator enums', async () => {
  const schemaPath = path.resolve(__dirname, '../mocks/enums.yaml')
  const oas = await new OasManager().parse(schemaPath)
  const defaultGenerator = new SchemaGenerator(
    {
      usedEnumNames: {},
      enumType: 'asConst',
      enumSuffix: '',
      dateType: 'string',
      optionalType: 'questionToken',
      transformers: {},
      oasType: false,
      unknownType: 'any',
    },
    {
      oas,
      pluginManager: mockedPluginManager,
      plugin: {} as Plugin<PluginOptions>,
      contentType: undefined,
      include: undefined,
    },
  )

  const schemas = oas.getDefinition().components?.schemas

  test('generate x-enum-varnames Type', async () => {
    const node = defaultGenerator.buildSource('enumVarNames', schemas?.['enumVarNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate x-enumNames Type', async () => {
    const node = defaultGenerator.buildSource('enumNames', schemas?.['enumNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate enum Items', async () => {
    const node = defaultGenerator.buildSource('enumItems', schemas?.['enum.Items'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate enum String', async () => {
    const node = defaultGenerator.buildSource('enumString', schemas?.['enum.String'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate enum AllOf', async () => {
    const node = defaultGenerator.buildSource('enumObject', schemas?.['enum.AllOf'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate enum InObject', async () => {
    const node = defaultGenerator.buildSource('enumObject', schemas?.['enum.InObject'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate enum Array', async () => {
    const node = defaultGenerator.buildSource('enumArray', schemas?.['enum.Array'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate with enumtype enum', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'enum',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as OasTypes.SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate with enumtype asPascalConst', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'asPascalConst',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as OasTypes.SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate with enumtype constEnum', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'constEnum',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })

  test('generate with enumtype literal', async () => {
    const generator = new SchemaGenerator(
      {
        usedEnumNames: {},
        enumType: 'literal',
        enumSuffix: '',
        dateType: 'string',
        optionalType: 'questionToken',
        transformers: {},
        oasType: false,
        unknownType: 'any',
      },
      {
        oas,
        pluginManager: mockedPluginManager,
        plugin: {} as Plugin<PluginOptions>,
        contentType: undefined,
        include: undefined,
      },
    )

    const node = generator.buildSource('enumNames', schemas?.['enumNames.Type'] as SchemaObject)

    expect(node).toMatchSnapshot()
  })
})
