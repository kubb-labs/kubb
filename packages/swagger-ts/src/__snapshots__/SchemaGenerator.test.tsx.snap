// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SchemaGenerator discriminator > Cat.type defined as const 1`] = `
[
  "export type Cat = {
    /**
     * @type string uuid
    */
    id: string;
    type: "Cat";
    /**
     * @type boolean | undefined
    */
    meow?: boolean;
};
",
]
`;

exports[`SchemaGenerator discriminator > Dog.type defined as const 1`] = `
[
  "export type Dog = {
    /**
     * @type string uuid
    */
    id: string;
    /**
     * @type number | undefined
    */
    bark?: number;
    type: "Dog";
};
",
]
`;

exports[`SchemaGenerator discriminator > MixedValueTypeConst ignores type constraint in favor of const constraint 1`] = `
[
  "export type MixedValueTypeConst = {
    foobar?: "foobar";
};
",
]
`;

exports[`SchemaGenerator discriminator > NullConst correctly produces "null" 1`] = `
[
  "export type NullConst = {
    withoutValue?: null;
};
",
]
`;

exports[`SchemaGenerator discriminator > NumberValueConst const correctly produces \`42\` 1`] = `
[
  "export type NumberValueConst = {
    foobar?: 42;
};
",
]
`;

exports[`SchemaGenerator discriminator > PetStore defined as array with type union 1`] = `
[
  "export type Petstore = {
    /**
     * @type integer
    */
    id: number;
    /**
     * @type array | undefined
    */
    pets?: (Cat | Dog)[];
};
",
]
`;

exports[`SchemaGenerator discriminator > StringValueConst const correctly produces "foobar" 1`] = `
[
  "export type StringValueConst = {
    foobar?: "foobar";
};
",
]
`;

exports[`SchemaGenerator enums > generate enum AllOf 1`] = `
[
  "export const enumObjectReason = {
    "created_at": "created_at",
    "description": "description"
} as const;
export type EnumObjectReason = (typeof enumObjectReason)[keyof typeof enumObjectReason];
export type enumObject = {
    reason?: EnumObjectReason;
};
",
]
`;

exports[`SchemaGenerator enums > generate enum Array 1`] = `
[
  "export const enumArrayIdentifier = {
    "NW": "NW",
    "NE": "NE",
    "SW": "SW",
    "SE": "SE"
} as const;
export type EnumArrayIdentifier = (typeof enumArrayIdentifier)[keyof typeof enumArrayIdentifier];
export type enumArray = {
    identifier?: [
        number,
        string,
        EnumArrayIdentifier
    ];
};
",
]
`;

exports[`SchemaGenerator enums > generate enum InObject 1`] = `
[
  "export const enumObjectReason2 = {
    "created_at": "created_at",
    "description": "description"
} as const;
export type EnumObjectReason2 = (typeof enumObjectReason2)[keyof typeof enumObjectReason2];
export type enumObject = {
    /**
     * @type string | undefined
    */
    reason?: EnumObjectReason2;
};
",
]
`;

exports[`SchemaGenerator enums > generate enum Items 1`] = `
[
  "export const enumItems = {
    "created_at": "created_at",
    "description": "description"
} as const;
export type EnumItems = (typeof enumItems)[keyof typeof enumItems];
export type enumItems = EnumItems[];
",
]
`;

exports[`SchemaGenerator enums > generate enum String 1`] = `
[
  "export const enumString = {
    "created_at": "created_at",
    "description": "description",
    "FILE.UPLOADED": "FILE.UPLOADED",
    "FILE.DOWNLOADED": "FILE.DOWNLOADED"
} as const;
export type EnumString = (typeof enumString)[keyof typeof enumString];
export type enumString = EnumString;
",
]
`;

exports[`SchemaGenerator enums > generate with enumtype asPascalConst 1`] = `
[
  "export const EnumNames = {
    "Pending": 0,
    "Received": 1
} as const;
export type EnumNames = (typeof EnumNames)[keyof typeof EnumNames];
export type enumNames = EnumNames;
",
]
`;

exports[`SchemaGenerator enums > generate with enumtype constEnum 1`] = `
[
  "export const enum EnumNames {
    "Pending" = 0,
    "Received" = 1
}
export type enumNames = EnumNames;
",
]
`;

exports[`SchemaGenerator enums > generate with enumtype enum 1`] = `
[
  "export enum EnumNames {
    "Pending" = 0,
    "Received" = 1
}
export type enumNames = EnumNames;
",
]
`;

exports[`SchemaGenerator enums > generate with enumtype literal 1`] = `
[
  "export type EnumNames = 0 | 1;
export type enumNames = EnumNames;
",
]
`;

exports[`SchemaGenerator enums > generate x-enum-varnames Type 1`] = `
[
  "export const enumVarNames = {
    "Pending": 0,
    "Received": 1
} as const;
export type EnumVarNames = (typeof enumVarNames)[keyof typeof enumVarNames];
export type enumVarNames = EnumVarNames;
",
]
`;

exports[`SchemaGenerator enums > generate x-enumNames Type 1`] = `
[
  "export const enumNames = {
    "Pending": 0,
    "Received": 1
} as const;
export type EnumNames = (typeof enumNames)[keyof typeof enumNames];
export type enumNames = EnumNames;
",
]
`;

exports[`SchemaGenerator petStoreRef > generate type for Pets 1`] = `
[
  "export type Pets = Pet[];
",
]
`;

exports[`TypeScript SchemaGenerator petStore > generate type for Pet with optionalType \`questionToken\` 1`] = `
[
  "export type Pet = {
    /**
     * @type integer int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string;
};
",
]
`;

exports[`TypeScript SchemaGenerator petStore > generate type for Pet with optionalType \`questionTokenAndUndefined\` 1`] = `
[
  "export type Pet = {
    /**
     * @type integer int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string | undefined;
};
",
]
`;

exports[`TypeScript SchemaGenerator petStore > generate type for Pet with optionalType \`undefined\` 1`] = `
[
  "export type Pet = {
    /**
     * @type integer int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag: string | undefined;
};
",
]
`;

exports[`TypeScript SchemaGenerator petStore > generate type for Pets 1`] = `
[
  "export type Pets = {
    /**
     * @type integer int64
    */
    id: number;
    /**
     * @type string
    */
    name: string;
    /**
     * @type string | undefined
    */
    tag?: string;
}[];
",
]
`;

exports[`TypeScript SchemaGenerator petStore > generate type for nullable fields 1`] = `
[
  "export type Test = {
    /**
     * @type string
    */
    foo: string | null;
};
",
]
`;
